library/math/num.lisp

```diff
@@ -79,139 +88,137 @@
-      (lisp Integer (a b) (cl:* a b)))
-    (define (fromInt x)
-      x))
-
-  (define-instance (Num I8)
-    (define (+ a b)
-      (lisp I8 (a b)
-        (%handle-8bit-overflow (cl:+ a b))))
-    (define (- a b)
-      (lisp I8 (a b)
-        (%handle-8bit-overflow (cl:- a b))))
-    (define (* a b)
-      (lisp I8 (a b)
-        (%handle-8bit-overflow (cl:* a b))))
-    (define (fromInt x)
-      (lisp I8 (x)
-        (%handle-8bit-overflow x))))
-
-  (define-instance (Num I16)
-    (define (+ a b)
-      (lisp I16 (a b)
-        (%handle-16bit-overflow (cl:+ a b))))
-    (define (- a b)
-      (lisp I16 (a b)
-        (%handle-16bit-overflow (cl:- a b))))
-    (define (* a b)
-      (lisp I16 (a b)
-        (%handle-16bit-overflow (cl:* a b))))
-    (define (fromInt x)
-      (lisp I16 (x)
-        (%handle-16bit-overflow x))))
-
-  (define-instance (Num I32)
-    (define (+ a b)
-      (lisp I32 (a b)
-        (%handle-32bit-overflow (cl:+ a b))))
-    (define (- a b)
-      (lisp I32 (a b)
-        (%handle-32bit-overflow (cl:- a b))))
-    (define (* a b)
-      (lisp I32 (a b)
-        (%handle-32bit-overflow (cl:* a b))))
-    (define (fromInt x)
-      (lisp I32 (x)
-        (%handle-32bit-overflow x))))
-
-  (define-instance (Num I64)
-    (define (+ a b)
-      (lisp I64 (a b)
-        (%handle-64bit-overflow (cl:+ a b))))
-    (define (- a b)
-      (lisp I64 (a b)
-        (%handle-64bit-overflow (cl:- a b))))
-    (define (* a b)
-      (lisp I64 (a b)
-        (%handle-64bit-overflow (cl:* a b))))
-    (define (fromInt x)
-      (lisp I64 (x)
-        (%handle-64bit-overflow x))))
-
-  (define-instance (Num IFix)
-    (define (+ a b)
-      (lisp IFix (a b)
-        (cl:+ a b)))
-    (define (- a b)
-      (lisp IFix (a b)
-        (cl:- a b)))
-    (define (* a b)
-      (lisp IFix (a b)
-        (cl:* a b)))
-    (define (fromInt x)
-      (lisp IFix (x)
-        (%handle-fixnum-overflow x)))))
-
-
-(cl:defmacro %define-unsigned-num-instance (coalton-type bits)
-  `(coalton-toplevel
-     (define-instance (Num ,coalton-type)
+  (define-ord Integer)
+  (define-ord IFix)
+  (define-ord UFix)
+  (define-ord I8)
+  (define-ord U8)
+  (define-ord I16)
+  (define-ord U16)
+  (define-ord I32)
+  (define-ord U32)
+  (define-ord I64)
+  (define-ord U64)
+  (define-ord Single-Float)
+  (define-ord Double-Float))
+
+;;;
+;;; Overflow checks for signed values
+;;;
+
+(cl:declaim (cl:inline %unsigned->signed))
+(cl:defun %unsigned->signed (bits x)
+  ;; This is the two's complement conversion of X (interpreted as BITS
+  ;; bits) to a signed integer (as a Lisp object).
+  (cl:-
+   (cl:ldb (cl:byte (cl:1- bits) 0) x)
+   (cl:dpb 0 (cl:byte (cl:1- bits) 0) x)))
+
+(cl:defmacro %define-overflow-handler (name bits)
+  `(cl:progn
+     (cl:declaim (cl:inline ,name))
+     (cl:defun ,name (value)
+       (cl:typecase value
+         ((cl:signed-byte ,bits) value)
+         (cl:otherwise
+          (cl:cerror "Continue, wrapping around."
```

Comment:
Any consideration to, in release mode, eliding the test and the cerror, and just wrapping unconditionally?

Reply:
I think that would just hide bugs.